mmcv/runner/base_module.py

```diff
@@ -38,26 +83,99 @@ def is_init(self):
 
     def init_weights(self):
         """Initialize the weights."""
-        from ..cnn import initialize
 
+        # check if it is top-level module
+        is_top_level_module = len(self._params_init_info) == 0
+        if is_top_level_module:
+            # Initialize the `_params_init_info`,
+            # When detecting the `tmp_mean_value` of
+            # the corresponding parameter is changed, update related
+            # initialization information
+            for name, param in self.named_parameters():
+                self._params_init_info[param]['param_name'] = name
+                self._params_init_info[param][
+                    'init_info'] = f'The value is the same before and ' \
+                                   f'after calling `init_weights` ' \
+                                   f'of {self.__class__.__name__} '
+                self._params_init_info[param][
+                    'tmp_mean_value'] = param.data.mean()
+
+            # pass `params_init_info` to all submodules
+            # All submodules share the same `params_init_info`,
+            # so it will be updated when parameters are
+            # modified at any level of the model.
+            for sub_module in self.modules():
+                sub_module._params_init_info = self._params_init_info
+
+        # Get the initialized logger, if not exist,
+        # create a logger named `mmcv`
+        logger_names = list(logger_initialized.keys())
+        logger_name = logger_names[0] if logger_names else 'mmcv'
+
+        from ..cnn import initialize
+        module_name = self.__class__.__name__
         if not self._is_init:
             if self.init_cfg:
+                print_log(
+                    f'initialize {module_name} with init_cfg {self.init_cfg}',
+                    logger=logger_name)
                 initialize(self, self.init_cfg)
-                if isinstance(self.init_cfg, (dict, ConfigDict)):
-                    # Avoid the parameters of the pre-training model
-                    # being overwritten by the init_weights
+                if isinstance(self.init_cfg, dict):
+                    # Avoid the parameters of the pretrained model
+                    # are overwritten by the `init_weights`
                     # of the children.
                     if self.init_cfg['type'] == 'Pretrained':
                         return
 
             for m in self.children():
                 if hasattr(m, 'init_weights'):
                     m.init_weights()
+                    # users may overload the `init_weights`
```

Comment:
If the children modules are inherited from `BaseModule`, how do you decide whether it is user-defined `init_weights`?

Reply:
If children modules are initialized by the `init_cfg`  after `m.init_weights()`, the `temp_mean` and `init_info` should have been updated, so `temp_mean` is consistent with the parameters, so it would not be updated in the next line.